parlai/agents/transformer/polyencoder.py

```diff
@@ -330,5 +359,19 @@ def forward(
         elif (
             ctxt_rep is not None and ctxt_rep_mask is not None and cand_rep is not None
         ):
-            return self.score(ctxt_rep, ctxt_rep_mask, cand_rep)
+            # ctxt_pos can be none, if we are using codes (not first M)
+            return self.score(ctxt_rep, ctxt_rep_mask, ctxt_pos, cand_rep)
         raise Exception('Unsupported operation')
+
+
+class PolyBasicAttention(BasicAttention):
+    def __init__(self, poly_type, n_codes, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.poly_type = poly_type
+        self.n_codes = n_codes
+
+    def forward(self, *args, **kwargs):
+        lhs_emb = super().forward(*args, **kwargs)
+        if self.poly_type == 'codes' and self.n_codes == 1 and len(lhs_emb.shape) == 2:
+            lhs_emb = lhs_emb.unsqueeze(self.dim - 1)
```

Comment:
can you explain this?

Reply:
This is a bit of a tricky situation. When we are using the `codes` polyencoder, the output of the first level of attention is `bsz x n_codes x hsz`; since there is only 1 code, the `BasicAttention` squeezes to make it `bsz x hsz`; thus, we need to unsqueeze to have the appropriate dimensionality for use later in the pipeline.